Custom mircroservice tutorial dotnet version (#165)
mia-platform/documentation: docs/tutorial/node_ms/node_ms_tutorial.mdx

</p>
</details>

</TabItem>
<TabItem value="dotnet">

The [BrewUp .NET Example](https://github.com/BrewUp/BrewUpApiTemplate) is based on the [Minimal API](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis/overview?view=aspnetcore-6.0) approach. We will also use [xunit.net](https://xunit.net/) for testing and [moq](https://github.com/moq/moq4) for mocking.

We can create a new xunit test project in the solution with the following commands:
``` bash
cd src/
dotnet new xunit -o Tests
dotnet sln BrewUpApiTemplate.sln add Tests/Tests.csproj
```

We have to add the following dependencies for testing and mocking as well a reference to the main project:
``` bash
cd Tests/
dotnet add package Microsoft.AspNetCore.Mvc.Testing --version 6.0.13
dotnet add package Moq --version 4.18.4
dotnet add package Moq.Contrib.HttpClient --version 1.4.0
dotnet add reference ../BrewUpApiTemplate/BrewUpApiTemplate.csproj
```

We can now start to develop the tests.

:::info

> Disclamer: the current example doesn't follow the .NET Framework best practice for simplicity of demonstration. Feel free to refactor the results as you see fit.

Reviewer: ```suggestion
Disclamer: the current example does not follow the .NET Framework best practice for simplicity of demonstration. Feel free to refactor the results as you see fit.
```

:::

> In order for the test framework to reach our web application we have to to add the following line at the end of the `src/BrewUpApiTemplate/Program.cs` file.

Reviewer: ```suggestion
In order for the test framework to reach our web application we have to add the following line at the end of the `src/BrewUpApiTemplate/Program.cs` file.
```

        this._id = id;
        this.name = name;
        this.surname = surname;
        this.email = email;
        this.phone = phone;
    }

    public string _id { get; set; }
    public string name { get; set; }
    public string surname { get; set; }
    public string email { get; set; }
    public string phone { get; set; }
}
```

``` csharp title=BrewUpApiTemplate/Models/Order.cs
namespace BrewUpApiTemplate.Models;

public class Order
{
    public Order(string customerId) {
        this.customerId = customerId;
    }

    public string customerId { get; set; }
}
```

``` csharp title=BrewUpApiTemplate/Models/ServiceResponse.cs
namespace BrewUpApiTemplate.Models;

public class ServiceResponse
{
    public ServiceResponse(double shippingCost) {
        this.shippingCost = shippingCost;
    }

    public double shippingCost { get; set; }
}
```

``` csharp title=BrewUpApiTemplate/Models/ServiceErrorResponse.cs
namespace BrewUpApiTemplate.Models;

public class ServiceErrorResponse
{
    public ServiceErrorResponse(string error) {
        this.error = error;
    }

    public string error { get; set; }
}
```

> We have to create a mock for the web application we will use so that we can test it without making the real service requests. Create a `MockWebApplicationFactory.cs` file on the root of the **Tests** project with the following content.

Reviewer: ```suggestion
We have to create a mock for the web application we will use in order to test it without making the real service requests. Create a `MockWebApplicationFactory.cs` file on the root of the **Tests** project with the following content.
```

using BrewUpApiTemplate.Models;
using Moq;
using System.Net.Http;
using Microsoft.AspNetCore.TestHost;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Options;
using Moq.Contrib.HttpClient;
using System.Collections.Generic;
using Microsoft.Extensions.Configuration;
using System.Threading.Tasks;
using System;
using System.Net;

namespace Tests;

public class UnitTest1: IClassFixture<MockWebApplicationFactory>
{
    private readonly double DEFAULT_SHIPPING_COST = 10;
    private readonly double NEW_CUSTOMER_SHIPPING_COST = 5.99;
    private HttpClient client;

    public UnitTest1(MockWebApplicationFactory factory)
    {
        this.client = factory.CreateClient();
    }

    [Fact]
    public async void TestNewCustomerShippingCost()
    {
        var httpResponse = await client.GetAsync($"/shipping-cost/?customerEmail=test-email-new");

        Assert.True(httpResponse.IsSuccessStatusCode);

        var response = await httpResponse.Content.ReadAsStringAsync();
        var serviceResponse = JsonConvert.DeserializeObject<ServiceResponse>(response);

        Assert.Equal(NEW_CUSTOMER_SHIPPING_COST, serviceResponse?.shippingCost);
    }
}
```

> In this test we tested only the "happy path," that is, returning the correct shipping cost for a new customer. To achieve full coverage by testing the inputs and the return of the default shipping cost you need to write more tests, so if you want to try that or just copy the complete test code below:

Reviewer: ```suggestion
In this test, we only tested the "happy path", i.e. the return of the correct shipping cost for a new customer. To achieve full coverage by testing the inputs and the return of the default shipping cost you need to write more tests, so if you want to try that or just copy the complete test code below:
```
